File: java to c++ encoders/TFTPErrorEncoder.cpp
#include "TFTPErrorEncoder.h"
#include "EncoderUtils.h"

namespace bgu
{
	namespace spl171
	{
		namespace net
		{
			namespace impl
			{
				namespace TFTP
				{
					namespace CommandEncoders
					{
						using bgu::spl171::net::impl::TFTP::Commands::Command;
						using bgu::spl171::net::impl::TFTP::Commands::Responses::TFTPError;
						using bgu::spl171::net::impl::TFTP::EncoderDecoders::ShortEncoderDecoder;
						using bgu::spl171::net::impl::TFTP::EncoderDecoders::StringEncoderDecoder;

						std::vector<char> TFTPErrorEncoder::encode(Command *cmd)
						{
							TFTPError *error = static_cast<TFTPError*>(cmd);
							std::vector<char> opCode = shortEncoderDecoder->encode(error->getOpCode());
							std::vector<char> errorCode = shortEncoderDecoder->encode(error->getErrorCode());
							std::vector<char> errorMsg = stringEncoderDecoder->encode(error->errorMessage());
							return EncoderUtils::mergeByteArray({opCode, errorCode, errorMsg});
						}
					}
				}
			}
		}
	}
}

File: Server/src/main/java/bgu/spl171/net/impl/TFTP/CommandEncoders/TFTPErrorEncoder.java
package bgu.spl171.net.impl.TFTP.CommandEncoders;

import bgu.spl171.net.impl.TFTP.Commands.Command;
import bgu.spl171.net.impl.TFTP.Commands.Responses.TFTPError;
import bgu.spl171.net.impl.TFTP.EncoderDecoders.ShortEncoderDecoder;
import bgu.spl171.net.impl.TFTP.EncoderDecoders.StringEncoderDecoder;

public class TFTPErrorEncoder implements CommandEncoder{
    private ShortEncoderDecoder shortEncoderDecoder = new ShortEncoderDecoder();
    private StringEncoderDecoder stringEncoderDecoder = new StringEncoderDecoder();

    @Override
    public byte[] encode(Command cmd) {
        TFTPError error = (TFTPError) cmd;
        byte[] opCode = shortEncoderDecoder.encode(error.getOpCode());
        byte[] errorCode = shortEncoderDecoder.encode(error.getErrorCode());
        byte[] errorMsg = stringEncoderDecoder.encode(error.getErrorMessage());
        return EncoderUtils.mergeByteArray(opCode,errorCode,errorMsg);
    }
}
